### ISSUE TITLE
Pathed Result does not support implicit conversion for IEnumerables

### ISSUE THREAD

**Issue by brandonmhall** [tombatron/NReJSON#24]
When calling the `GetJson<T>` method, if you set `T` to `IEnumerable<TI>` implicit conversion fails. This feels like incorrect behaviour.

The below exception is thrown during deserialisation:

```
System.Text.Json.JsonException
  HResult=0x80131500
  Message=The JSON value could not be converted to System.Collections.Generic.IEnumerable`1[System.Object]. Path: $[0] | LineNumber: 0 | BytePositionInLine: 8.
  Source=System.Text.Json
  StackTrace:
   at System.Text.Json.ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(Type propertyType)
   at System.Text.Json.Serialization.JsonCollectionConverter`2.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, TCollection& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonCollectionConverter`2.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, TCollection& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 utf8Json, JsonTypeInfo jsonTypeInfo, Nullable`1 actualByteCount)
   at System.Text.Json.JsonSerializer.ReadFromSpan[TValue](ReadOnlySpan`1 json, JsonTypeInfo jsonTypeInfo)
   at System.Text.Json.JsonSerializer.Deserialize[TValue](String json, JsonSerializerOptions options)
   at Flow.Runtime.RedisInterface.DefaultSerializer.Deserialize[TResult](RedisResult serializedValue) in C:\projects\Flow Core\Runtime\Flow.Runtime.Redis\Serializer.cs:line 17
   at NReJSON.PathedResult`1.ParseRedisResult()
   at NReJSON.PathedResult`1.GetEnumerator()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Flow.Runtime.RedisInterface.Repository.RedisRepository.<GetJsonArray>d__8`1.MoveNext() in C:\projects\Flow Core\Runtime\Flow.Runtime.Redis\Repository\RedisRepository.cs:line 97

  This exception was originally thrown at this call stack:
    [External Code]
    Flow.Runtime.RedisInterface.DefaultSerializer.Deserialize<TResult>(StackExchange.Redis.RedisResult) in Serializer.cs
    [External Code]
    Flow.Runtime.RedisInterface.Repository.RedisRepository.GetJsonArray<T>(string, string) in RedisRepository.cs
```

This feels like incorrect behavior when getting json

**Comment by tombatron** [comment]
Hey there!

Could you provide me with some sample JSON and a code snippet so I can reproduce the issue on my end please?

**Comment by brandonmhall** [comment]
Hey, sure:

code snippet from the package to interface against our redis db:

```c#
public async Task<T?> GetJson<T>(string key, string path = ".")
{
    try
    {
        return await this.database.JsonGetAsync<T>(key, path);
    }
    catch (Exception e)
    {
        this.logger.LogError(e, "Failed to get json for key {key}", key);
        return default;
    }
}
```

In the code to call that method:

```c#
List<Variable> variables = await this.repository.GetJson<List<Variable>>(key, ".Variables")

```

Where the json is along the lines of:

```json
{
  "Variables": [
    {
      "prop": "value"
    }
  ]
}
```

As a note on the above, I am returning `T` from the  method in the package. The code does work If i return a `PathedResult<T>` and set `T` to `Variable`, and assign it to `IEnumerable<T>`, but it would be preferable to be able to assign back the response directly to `T` - code snipped below if i change the return type to be Pathed

```c#
IEnumerable<Variable> variables = await this.repository.GetJson<Variable>(key, ".Variables")
```

**Comment by tombatron** [comment]
OK I think I see what's going on here. 

If I remember correctly, I introduced `PathedResult` to handle the a result from `JSON.GET` that was either a single object or an array. 

For example, using the sample JSON that you provided let's assume we load the following object:

`"{\"Variables\":[{\"prop\":\"value\"},{\"prop\":\"value2\"}]}"`

Note, the `Variables` array contains two objects. 

Issuing the command with the JSONPath that you provided:

`json.get issue-24 .Variables`

Would yield this result:

`"[{\"prop\":\"value\"},{\"prop\":\"value2\"}]"`

You can see that it yield a JSONArray that we could deserialize as an IEnumerable. 

However, if we get a little more specific with our JSONPath, and executed a command like:

`json.get issue-24 .Variables[0]`

We'll get the following result:

`"{\"prop\":\"value\"}"`

The `PathedResult` class is intended to handle both conditions. If you check out the source, the `PathedResult` class implements `IEnumerable<TResult>`, but if you look further, it also implements an implicit conversion to `TResult`. 

Long story short, the typed version of `JsonGetAsync` basically assumes that the result is going to be an array, but has support for treating the result as a single value. 

I would look at changing the type that your `GetJson` method returns so that it returns an `IEnumerable<T?>` and only call that method with the type represented within your array.

**Comment by brandonmhall** [comment]
Thanks for getting back to me so quickly :)

Do you think there would be value in allowing implicit conversion to check the type of `T` and return the response accordingly? For example check if `T is IEnumerable` return the full result as an IEnumerable, otherwise return `pathedResult.Single()`

In order to handle both cases I have added a `GetJsonArray` method which returns `IEnumerable<T>` which works you suggested.

**Comment by tombatron** [comment]
> Do you think there would be value in allowing implicit conversion to check the type of `T` and return the response accordingly?

I'd be happy to take a look at a PR if you'd like to propose how. 😊
